<a href="https://colab.research.google.com/github/Amit198181/JBVNL/blob/main/MDI_zero_All_Months_100Units.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
import os

uploaded = files.upload()

for fn in uploaded.keys():
    print('User uploaded file "{name}" with length {length} bytes'.format(
        name=fn, length=len(uploaded[fn])))

Saving LTIS_APR_AUG.xlsx to LTIS_APR_AUG.xlsx
User uploaded file "LTIS_APR_AUG.xlsx" with length 2230734 bytes


In [2]:
import pandas as pd

file_name = next(iter(uploaded))
df = pd.read_excel(file_name)

display(df.head())

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

  CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
0         URBAN  ...                    0.0                      0.0   
1         RURAL  ...                    0.0                      0.0   
2         RURAL  ...                    0.0                      0.0   
3         RURAL  ...                    0.0                      0.0   
4         RURAL  ...                    0.0                      0.0   

  BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
0                    95.0           12.0                     0.0   
1                    30.0            5.0                     0.0   
2                    61.0            6.0                     0.0   
3                    55.0            4.0                     0.0   
4                    25.0            5.0                     0.0   

   SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
0                      0.0                    275.0            18.0   
1                      0.0                     70.0             4.0   
2                      0.0                    120.0             5.0   
3                      0.0                    150.0             6.0   
4                      0.0                    102.0             6.0   

   LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
0                     0.0                        0  
1                     0.0                        0  
2                     0.0                        0  
3                     0.0                        0  
4                     0.0                        0  

[5 rows x 30 columns]

Identify mdi columns: Find all columns in the DataFrame that represent the maximum demand (MDI) for each month.
Check for zero mdi across months: Create a condition to check if the value in all identified MDI columns is zero for each row.
Filter dataframe: Apply the condition to filter the DataFrame and keep only the rows where MDI is zero across all months.
Display results: Show the filtered DataFrame to the user.
Finish task: Explain the results and suggest potential next steps.

# Task
Analyze the dataframe `df` to find rows where the maximum demand (MDI) is zero for all months.

## Identify mdi columns

### Subtask:
Find all columns in the DataFrame that represent the maximum demand (MDI) for each month.


**Reasoning**:
Identify all columns in the DataFrame that contain 'MDI' in their name.



In [3]:
mdi_columns = [col for col in df.columns if 'MDI' in col]
print("MDI columns:", mdi_columns)

MDI columns: ['MDI_HTS_202504', 'MDI_HTS_202505', 'MDI_HTS_202506', 'MDI_HTS_202507', 'MDI_HTS_202508']


## Check for zero mdi across months

### Subtask:
Create a condition to check if the value in all identified MDI columns is zero for each row.


**Reasoning**:
Create a boolean condition to check if all values in the identified MDI columns are zero for each row, handling potential NaN values by treating them as non-zero for this condition.



In [4]:
zero_mdi_condition = (df[mdi_columns].fillna(-1) == 0).all(axis=1)

## Filter dataframe

### Subtask:
Apply the condition to filter the DataFrame and keep only the rows where MDI is zero across all months.


**Reasoning**:
Apply the previously created boolean condition `zero_mdi_condition` to filter the DataFrame `df` and store the result in a new DataFrame `df_zero_mdi`.



In [5]:
df_zero_mdi = df[zero_mdi_condition]

## Display results

### Subtask:
Show the filtered DataFrame to the user.


**Reasoning**:
Display the first few rows of the filtered DataFrame and print the number of rows.



In [6]:
display(df_zero_mdi.head())
print(f"Number of rows with zero MDI across all months: {len(df_zero_mdi)}")

ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
46   RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
67   RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
68   RANCHI  RANCHI  RANCHI EAST   M1821  11220768017            LIVE   
89   RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
309  RANCHI  RANCHI  RANCHI EAST  OM241G  11221068343            LIVE   

                        CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
46     ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02   
67     ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00   
68   ASSISTANT MECH ENG.PH.MECH.SUB          LTIS            20.00   
89          EXECUTIVE ENGINEER,PHED          LTIS             7.00   
309        THE DY.SUPDT.ELECT.ENGG.          LTIS            55.00   

    CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
46           GOVT  ...                    0.0                      0.0   
67         OTHERS  ...                    0.0                      0.0   
68         OTHERS  ...                    0.0                      0.0   
89         OTHERS  ...                    0.0                      0.0   
309        OTHERS  ...                    0.0                      0.0   

    BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
46                    404.0            0.0                     0.0   
67                   2938.0            0.0                     0.0   
68                     93.0            0.0                     0.0   
89                   1000.0            0.0                     0.0   
309                  5834.0            0.0                     0.0   

     SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
46                       0.0                   1213.0             0.0   
67                       0.0                   2693.0             0.0   
68                       0.0                     85.0             0.0   
89                       0.0                   1109.0             0.0   
309                      0.0                   7128.0             0.0   

     LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
46                      0.0                        0  
67                      0.0                        0  
68                      0.0                        0  
89                      0.0                        0  
309                     0.0                        0  

[5 rows x 30 columns]

Number of rows with zero MDI across all months: 368


## Summary:

### Data Analysis Key Findings

*   The columns identified as representing Maximum Demand (MDI) for each month are 'MDI\_HTS\_202504', 'MDI\_HTS\_202505', 'MDI\_HTS\_202506', 'MDI\_HTS\_202507', and 'MDI\_HTS\_202508'.
*   There are 368 rows in the DataFrame where the MDI is zero across all identified month columns.

### Insights or Next Steps

*   Investigate why these 368 rows have zero MDI across all months; this could indicate inactive accounts, data errors, or a specific customer segment.
*   Exclude these rows from further analysis if the focus is on customers with active demand.


In [13]:
from google.colab import files

# Save the filtered DataFrame to an Excel file
output_filename = 'MDI_Zero.xlsx'
df_zero_mdi.to_excel(output_filename, index=False)

# Provide a link to download the file
files.download(output_filename)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Summary:

### Data Analysis Key Findings

*   Columns representing energy consumption were identified as \['BILLED\_UNITS\_HTS\_202504', 'BILLED\_UNITS\_HTS\_202505', 'BILLED\_UNITS\_HTS\_202506', 'BILLED\_UNITS\_HTS\_202507', 'BILLED\_UNITS\_HTS\_202508'].
*   A boolean condition was created to identify rows where at least one of these consumption columns has a value of 100 or more, treating missing values as 0.
*   The DataFrame was filtered based on this condition.
*   The resulting filtered DataFrame contains 16297 rows, representing consumers with an energy consumption of 100 units or more in at least one month.

### Insights or Next Steps

*   The filtered dataset can now be used for further analysis specifically on high-consumption consumers.
*   Investigate the characteristics of these high-consumption consumers to identify patterns or reasons for their higher energy usage.


In [ ]:
# Identify the column for August's energy consumption
august_consumption_column = 'BILLED_UNITS_HTS_202508'

# Create a condition to check for consumption of 100 or more in August, handling potential NaN values
august_consumption_condition = df_filtered_consumption[august_consumption_column].fillna(0) >= 100

# Filter the DataFrame based on the condition
df_august_consumption = df_filtered_consumption[august_consumption_condition]

# Display the filtered DataFrame and the number of rows
display(df_august_consumption.head())
print(f"Number of consumers with 100+ units consumption in August: {len(df_august_consumption)}")